pkg/scheduling/volumes.go

```diff
@@ -0,0 +1,218 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/samber/lo"
+	csitranslation "k8s.io/csi-translation-lib"
+	"k8s.io/csi-translation-lib/plugins"
+	"knative.dev/pkg/logging"
+
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/utils/atomic"
+)
+
+const (
+	IsDefaultStorageClassAnnotation = "storageclass.kubernetes.io/is-default-class"
+)
+
+// translator is a CSI Translator that translates in-tree plugin names to their out-of-tree CSI driver names
+var translator = csitranslation.New()
+
+// Cache the lookup of our default storage class, global atomic cache since the volume usage is per node
+var defaultStorageClass = atomic.NewCachedVariable[string](1 * time.Minute)
+
+// ResetDefaultStorageClass is intended to be called from unit tests to reset the default storage class
+func ResetDefaultStorageClass() {
+	defaultStorageClass.Reset()
+}
+
+type Volumes map[string]sets.Set[string]
+
+func (u Volumes) Add(provisioner string, pvcID string) {
+	existing, ok := u[provisioner]
+	if !ok {
+		existing = sets.New[string]()
+		u[provisioner] = existing
+	}
+	existing.Insert(pvcID)
+}
+
+func (u Volumes) Union(vol Volumes) Volumes {
+	cp := Volumes{}
+	for k, v := range u {
+		cp[k] = sets.New(sets.List(v)...)
+	}
+	for k, v := range vol {
+		existing, ok := cp[k]
+		if !ok {
+			existing = sets.New[string]()
+			cp[k] = existing
+		}
+		existing.Insert(sets.List(v)...)
+	}
+	return cp
+}
+
+func (u Volumes) Insert(volumes Volumes) {
+	for k, v := range volumes {
+		existing, ok := u[k]
+		if !ok {
+			existing = sets.New[string]()
+			u[k] = existing
+		}
+		existing.Insert(sets.List(v)...)
+	}
+}
+
+func (u Volumes) Copy() Volumes {
+	cp := Volumes{}
+	for k, v := range u {
+		cp[k] = sets.New(sets.List(v)...)
+	}
+	return cp
+}
+
+//nolint:gocyclo
+func GetVolumes(ctx context.Context, kubeClient client.Client, pod *v1.Pod) (Volumes, error) {
```

Comment:
Should this be tracked in Cluster State? Perhaps on UpdatePod?

Reply:
I think this is fine where it is. There are a few different places where we are getting the pod volumes and I can see the utility of this being a utils function that can grab the volumes generically for a given pod.